spring-integration-core/src/main/java/org/springframework/integration/util/MessagingMethodInvokerHelper.java

```diff
@@ -314,22 +314,26 @@ public T process(Message<?> message) throws Exception {
 	}
 
 	/*
-	 * If there's a single method, the content is JSON, the payload is a
-	 * String or byte[], the parameter doesn't match the payload,
-	 * and there is a Json Object Mapper on the CP,
-	 * convert.
+	 * If there's a single method, it is SpEL only, the content is JSON,
+	 * the payload is a String or byte[], the parameter doesn't match the payload,
+	 * and there is a Json Object Mapper on the CP, convert.
 	 */
-	private Message<?> possiblyConvert(Message<?> message) throws Exception {
-		if (this.handlerMethod != null && this.handlerMethod.getTargetParameterType() != null &&
-				this.jsonObjectMapper != null) {
-			Class<?> type = this.handlerMethod.getTargetParameterType();
+	private Message<?> possiblyConvert(Message<?> message) {
+		if (this.handlerMethod != null &&
+				this.handlerMethod.exclusiveMethodParameter != null &&
+				this.jsonObjectMapper != null &&
+				this.handlerMethod.spelOnly) {
```

Comment:
It's not just `spelOnly` it could be `this.useSpelInvoker` too - see `processInternal()`.

Furthermore, we might fall back to SpEL in `invokeHandlerMethod` before `spelOnly` is forced to true.

I think this might need to be moved to `invokeExpression` so the conversion is done, if necessary, whenever SpEL is involved, which would mean making `ParametersWrapper.message` non-`final`.

Reply:
Yes, sounds better.

Will take a look later.

Thanks